# Week 1 Checkpoint Assignment: Data Preprocessing with pandas

**Name:** ______________________________  
**Student ID:** _________________________  
**Date:** _______________________________

**Points:** 60  
**Suggested time:** 60–75 minutes

This is an individual end-of-session exercise. It assesses your ability to inspect, filter, clean, and summarize tabular data with pandas.

## Submission requirements

Submit both of the following:

1. Your completed notebook, renamed **Lastname_Firstname_Week1_Checkpoint.ipynb**.
2. The cleaned CSV created in Task 7, named **hepatitis_cleaned.csv**.

Before submitting:

- run the notebook from top to bottom;
- resolve all error messages;
- keep your outputs visible;
- answer every written-response prompt in a Markdown cell;
- do not modify the original dataset file.

You may consult pandas documentation and course notes. Your submitted code and explanations must be your own.

## Dataset and allowed tools

Use the hepatitis dataset supplied with the lesson. The notebook loads the local copy first and uses the public source only as a fallback.

You may use standard pandas and NumPy methods covered in class, including:

- <code>head</code>, <code>shape</code>, <code>info</code>, and <code>describe</code>;
- column selection and <code>loc</code>;
- Boolean filtering;
- <code>select_dtypes</code> and <code>value_counts</code>;
- <code>isna</code>, <code>dropna</code>, and <code>fillna</code>;
- <code>duplicated</code>, <code>drop_duplicates</code>, and <code>reset_index</code>;
- <code>groupby</code>, <code>sort_values</code>, and <code>to_csv</code>.

## Setup

Run this cell without changing it.

In [18]:
from pathlib import Path
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", 25)

local_path = Path("data/hepatitis.csv")
data_url = "https://raw.githubusercontent.com/datasets/hepatitis/master/data/hepatitis.csv"
source = local_path if local_path.exists() else data_url

df_raw = pd.read_csv(source)
print(f"Dataset loaded from: {source}")

Dataset loaded from: https://raw.githubusercontent.com/datasets/hepatitis/master/data/hepatitis.csv


## Task 1 — Confirm the dataset (4 points)

1. Display the first **seven** rows.
2. Store the dataset dimensions in a variable named <code>dataset_shape</code>.
3. Print a sentence reporting the number of rows and columns.

Your code must use the DataFrame rather than typing the dimensions manually.

In [19]:
# Task 1 solution
df_raw.head(7)
dataset_shape = df_raw.shape
print(f"The dataset has {dataset_shape[0]} rows, and has {dataset_shape[1]} columns.")

The dataset has 155 rows, and has 20 columns.


## Task 2 — Inspect structure and quality (8 points)

1. Display the DataFrame's concise structural summary.
2. Store the data type of every column in <code>column_types</code>.
3. Create <code>numeric_summary</code>, containing descriptive statistics for numeric columns.
4. In the Markdown cell below, identify:
   - one numeric variable;
   - one categorical variable;
   - one column that contains missing values.

Include evidence from your output.

In [29]:
# Task 2 solution
df_raw.info()
column_types = df_raw.dtypes.tolist()
numeric_columns = df_raw.select_dtypes(include="number").columns.tolist()
categorical_columns = df_raw.select_dtypes(exclude="number").columns.tolist()

numeric_summary = df_raw[numeric_columns].describe(include="all").T
print(categorical_columns)
numeric_summary

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155 entries, 0 to 154
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   age              155 non-null    int64  
 1   sex              155 non-null    object 
 2   steroid          154 non-null    object 
 3   antivirals       155 non-null    bool   
 4   fatigue          154 non-null    object 
 5   malaise          154 non-null    object 
 6   anorexia         154 non-null    object 
 7   liver_big        145 non-null    object 
 8   liver_firm       144 non-null    object 
 9   spleen_palpable  150 non-null    object 
 10  spiders          150 non-null    object 
 11  ascites          150 non-null    object 
 12  varices          150 non-null    object 
 13  bilirubin        149 non-null    float64
 14  alk_phosphate    126 non-null    float64
 15  sgot             151 non-null    float64
 16  albumin          139 non-null    float64
 17  protime         

,count,mean,std,min,25%,50%,75%,max
age,155.0,41.200000,12.565878,7.0,32.00,39.0,50.00,78.0
bilirubin,149.0,1.427517,1.212149,0.3,0.70,1.0,1.50,8.0
alk_phosphate,126.0,105.325397,51.508109,26.0,74.25,85.0,132.25,295.0
sgot,151.0,85.894040,89.650890,14.0,31.50,58.0,100.50,648.0
albumin,139.0,3.817266,0.651523,2.1,3.40,4.0,4.20,6.4
protime,88.0,61.852273,22.875244,0.0,46.00,61.0,76.25,100.0


### Task 2 written response

4. In the Markdown cell below, identify:
   - one numeric variable;
One numeric variable is age. The average age being just over 41, with a standard deviation of 12 1/2 years.
   - one categorical variable;
One categorical variable is antivirals. It is a boolean category, and I will refrain from making an assumption on the meaning of it.
   - one column that contains missing values.
One column that contains missing values is 'liver_big'. It has 10 null values.

## Task 3 — Select and filter data (10 points)

### Part A

Create <code>selected_columns</code> containing only these columns, in this order:

- age
- sex
- albumin
- bilirubin
- class

Display its first five rows.

### Part B

Create <code>filtered_patients</code> containing patients who:

- are younger than 30; **and**
- have an albumin value greater than or equal to 4.

Keep only <code>age</code>, <code>albumin</code>, and <code>class</code>. Sort the result by age from youngest to oldest.

In [31]:
# Task 3 solution
selected_columns = df_raw[['age', 'sex', 'albumin', 'bilirubin', 'class']]
df_raw.head(5)

filtered_patients = df_raw[(df_raw["age"] < 30) & (df_raw["albumin"] >= 4)]
filtered_patients[['age', 'albumin', 'class']].sort_values('age')

,age,albumin,class
110,7,4.2,live
148,20,4.0,live
20,22,4.2,live
41,23,4.6,live
80,23,4.8,live
36,23,4.1,live
70,24,4.3,live
116,24,4.1,live
135,25,4.5,live
105,25,6.4,live


### Task 3 written response

How many rows meet both filtering conditions? Explain how you obtained the answer.

The way I was able to obtain this answer is using the filter mechanism from the instruction notebook and then using two Boolean conditions conjoined with the AND symbol in Python to perform the conjoin operation. Then filtering using the sort values dot method to have the age filter applied without any additional parameter since it is ascending by default. 

## Task 4 — Explore categorical variables (8 points)

1. Create a list named <code>categorical_columns</code> using pandas data-type selection. Do not type the column names manually.
2. Create <code>steroid_counts</code> showing every value in the steroid column, including missing values.
3. Display the number of unique non-missing values in every categorical column.
4. Briefly explain why including missing values in a frequency table can be useful.

In [22]:
# Task 4 solution

categorical_columns = df_raw.select_dtypes(exclude="number").columns.tolist()
steroid_counts = df_raw["steroid"]
df_unique = df_raw[categorical_columns].nunique()
print(df_unique)



sex                2
steroid            2
antivirals         2
fatigue            2
malaise            2
anorexia           2
liver_big          2
liver_firm         2
spleen_palpable    2
spiders            2
ascites            2
varices            2
histology          2
class              2
dtype: int64


### Task 4 written response

The reason why you should include missing values is because you want to make sure there is awareness on how much of a variable was missing in your summary statistic evaluation. In this way, you can make sure not to come to a conclusion that an output may be derived from 100 observations when really you only have 40 or 50.

## Task 5 — Analyze missing values (12 points)

Create a DataFrame named <code>missing_summary</code> with:

- one row per original column;
- a column named <code>missing_count</code>;
- a column named <code>missing_percent</code>, rounded to one decimal place;
- rows sorted from highest to lowest missing percentage.

Then:

1. display only columns with at least one missing value;
2. store the name of the column with the most missing values in <code>highest_missing_column</code>;
3. calculate how many rows would remain after dropping every row containing at least one missing value, and store the result in <code>complete_row_count</code>.

In [23]:
# Task 5 solution
missing_summary = pd.DataFrame({
    'missing_count':df_raw.isnull().sum(),
    'missing_percent':(df_raw.isnull().sum() / len(df_raw)*100).round(1)
})

missing_summary = missing_summary.sort_values('missing_percent', ascending=False)
print(missing_summary[missing_summary['missing_count'] > 0])
highest_missing_column = missing_summary.index[0]
complete_row_count = len(df_raw.dropna())
total_row_count = len(df_raw)
print(complete_row_count)
print(total_row_count)

                 missing_count  missing_percent
protime                     67             43.2
alk_phosphate               29             18.7
albumin                     16             10.3
liver_firm                  11              7.1
liver_big                   10              6.5
bilirubin                    6              3.9
spleen_palpable              5              3.2
spiders                      5              3.2
varices                      5              3.2
ascites                      5              3.2
sgot                         4              2.6
steroid                      1              0.6
malaise                      1              0.6
fatigue                      1              0.6
anorexia                     1              0.6
80
155


### Task 5 written response

Would dropping every incomplete row be a reasonable default decision for this dataset? Support your answer with the number and percentage of rows that would be lost.

No, dropping every incomplete row here would be an unreasonable default decision. With this data set, after dropping all rows with at least one missing value, only 80 rows will remain out of the 155 in the original data set. That would mean losing 75 rows, which is 55% of the data. That is too much of a loss to justify. 

## Task 6 — Build a cleaned copy (12 points)

Create <code>df_clean</code> as a copy of <code>df_raw</code>. Do not change <code>df_raw</code>.

Apply all of these rules to <code>df_clean</code>:

1. Fill missing numeric values with each column's median.
2. Fill missing non-numeric values with each column's mode.
3. Remove duplicate rows.
4. Reset the index after duplicate removal.

Finally, display:

- total missing values before and after cleaning;
- duplicate row count after cleaning;
- the shape of the cleaned DataFrame.

In [24]:
# Task 6 solution
df_clean = df_raw.copy()

numeric_columns = df_clean.select_dtypes(include="number").columns
categorical_columns = df_clean.select_dtypes(exclude="number").columns

numeric_fill_values = df_clean[numeric_columns].median()
categorical_fill_values = df_clean[categorical_columns].mode().iloc[0]

df_clean[numeric_columns] = df_clean[numeric_columns].fillna(numeric_fill_values)
df_clean[categorical_columns] = df_clean[categorical_columns].fillna(
    categorical_fill_values
)

missing_values = int(df_raw.isna().sum().sum())

duplicate_count = df_clean.duplicated().sum()
df_clean = df_clean.drop_duplicates()

df_clean = df_clean.reset_index(drop=True)

missing_values_after = int(df_clean.isna().sum().sum())

print("Missing values:", missing_values)
print("Missing values after: ", missing_values_after)
print(f"Duplicate rows removed: {duplicate_count}")
print(f"Cleaned DataFrame shape: {df_clean.shape}")

Missing values: 167
Missing values after:  0
Duplicate rows removed: 0
Cleaned DataFrame shape: (155, 20)


/var/folders/fy/qvv9hxhx3bq64v6zwvv1kq2r0000gn/T/ipykernel_37163/1632772206.py:11: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_clean[categorical_columns] = df_clean[categorical_columns].fillna(


### Task 6 written response

Explain one advantage and one limitation of median/mode imputation.

The advantage of the imputation uh, for median is that it is really simple to do and doesn't take a lot of computation. And the data to do it is already available. The averaging or central tendency of your data is remained as well, which is very important. This applies for mode as well. 

For limitations, the imputation is artificial and therefore kind of sacrifices some fidelity of your data because you're adding artificial duplicate data to your data set.

## Task 7 — Validate, summarize, and export (6 points)

1. Use assertions to confirm that:
   - <code>df_clean</code> has no missing values;
   - <code>df_clean</code> has no duplicate rows;
   - every age is between 0 and 120 inclusive.
2. Create <code>class_summary</code> containing the count and percentage of each value in the class column. Round percentages to one decimal place.
3. Export <code>df_clean</code> to **hepatitis_cleaned.csv** without saving the index.
4. In the Markdown cell below, explain why retaining <code>df_raw</code> is important.

In [25]:
# Task 7 solution

assert df_clean.isna().sum().sum() == 0, "Missing values remain"
assert not df_clean.duplicated().any(), "Duplicate rows remain"
assert df_clean["age"].between(0, 120).all(), "Unexpected age detected"
assert set(df_clean["class"]).issubset({"live", "die"}), "Unexpected class label"

print("All validation checks passed.")

class_summary = pd.DataFrame({
    'count': df_clean['class'].value_counts(),
    'percentage': (df_clean['class'].value_counts() / len(df_clean) * 100).round(1)
})

print(class_summary)

df_clean.to_csv('hepatitis_cleaned.csv', index=False)
print("Exported to hepatitis_cleaned.csv")

All validation checks passed.
       count  percentage
class                   
live     123        79.4
die       32        20.6
Exported to hepatitis_cleaned.csv


### Task 7 written response

The reason why retaining the raw data frame is important is because the original data is important to understand any of the transformations we applied. This is for others and this is for my future self if I needed to clean the process in the future or to recreate it on a newer data set. Also, when the raw data frame is kept, statistics can be run after to determine the output and the impact of decisions made in the transformations and cleaning of the data. And this can help improve data pipelines and processes in the future for my future projects.

## Submission self-check

Run this cell after completing all tasks. It checks required variable names and submission files. Passing this check does not guarantee full credit; code quality and written explanations are also graded.

In [32]:
required_variables = [
    "dataset_shape",
    "column_types",
    "numeric_summary",
    "selected_columns",
    "filtered_patients",
    "categorical_columns",
    "steroid_counts",
    "missing_summary",
    "highest_missing_column",
    "complete_row_count",
    "df_clean",
    "class_summary",
]

missing_variables = [
    name for name in required_variables if name not in globals()
]

if missing_variables:
    print("Still missing required variables:")
    for name in missing_variables:
        print(" -", name)
else:
    print("All required variables exist.")

output_file = Path("hepatitis_cleaned.csv")
print("Cleaned CSV exists:", output_file.exists())

if "df_raw" in globals():
    print("Raw dataset still contains missing values:", df_raw.isna().sum().sum() > 0)

if "df_clean" in globals():
    print("Clean dataset has zero missing values:", df_clean.isna().sum().sum() == 0)
    print("Clean dataset has zero duplicates:", not df_clean.duplicated().any())

All required variables exist.
Cleaned CSV exists: True
Raw dataset still contains missing values: True
Clean dataset has zero missing values: True
Clean dataset has zero duplicates: True


## Optional challenge — Reusable function (no extra credit required)

Write a function named <code>summarize_missing(data)</code> that accepts any DataFrame and returns a missing-value summary with counts and percentages, sorted from most to least missing.

Test the function using <code>df_raw</code>.

In [12]:
# Optional challenge


## Grading rubric

| Area | Points | Full-credit evidence |
|---|---:|---|
| Task 1: Load and confirm | 4 | Correct preview, programmatic shape, and clear printed result |
| Task 2: Inspect | 8 | Appropriate inspection methods, stored results, evidence-based explanation |
| Task 3: Select and filter | 10 | Correct columns, Boolean conditions, ordering, and row count |
| Task 4: Categories | 8 | Programmatic type selection, counts including missing values, explanation |
| Task 5: Missingness | 12 | Accurate count/percentage table, sorting, complete-row calculation, justified decision |
| Task 6: Cleaning | 12 | Copy preserved, correct imputation, duplicate handling, before/after evidence |
| Task 7: Validate and export | 6 | Assertions pass, class summary is correct, CSV exported, explanation completed |
| **Total** | **60** | |

Code should be readable, use meaningful variable names, and avoid unnecessary repetition.